### PR TITLE
Implement class definition interface

### PR DESCRIPTION
**Pull request by kjkooistra-youwe**
Added interface for class definition, since the class is final.
This allows using the interface for proxies, like unit test mock objects.
Relates to #13595

### CODE REVIEW

**Comment by kjkooistra-youwe** [comment]
Tested:

- Classes editor, export and import
- Field collections
- Bricks
- Blocks
- Grid, grid options and favorites
- Classes rebuild

File: models/DataObject/ClassDefinitionInterface.php [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+
+    public function getCreationDate(): ?int;
+
+    public function getModificationDate(): ?int;
+
+    public function getUserOwner(): ?int;
+
+    public function getUserModification(): ?int;
+
+    public function setId(string $id): static;
```
Comment (blankse) [inline models/DataObject/ClassDefinitionInterface.php]: Can you add `@return $this` php docs to all methods which return $this?

Comment (kjkooistra-youwe) [reply models/DataObject/ClassDefinitionInterface.php]: Does that have any benefit when using typehints? The ClassDefinition class doesn't have it either.

Comment (blankse) [reply models/DataObject/ClassDefinitionInterface.php]: Yes there are benefits, see comments in https://github.com/pimcore/pimcore/pull/13214
Yes there are missing. I will add it in https://github.com/pimcore/pimcore/pull/14300

**Comment by brusch** [comment]
Related to #14337

**Comment by kjkooistra-youwe** [comment]
- Resolved conflict caused by https://github.com/pimcore/pimcore/commit/d505e46b8b9ff9505e2fb1f1bcb0616321a3b8a8
- Resolved errors caused by updated typehints in https://github.com/pimcore/pimcore/commit/b07426d048e81be7037582e77566ec59e2186b2e
- Resolved errors in definition import commands. Replaced `AbstractModel` with `ModelInterface`.

File: models/DataObject/ClassDefinitionInterface.php [hunk 2]
```diff
@@ -0,0 +1,306 @@
...
+    /**
+     * @internal
+     */
+    public function generateClassFiles(bool $generateDefinitionFile = true): void;
+
+    public function delete(): void;
+    /**
+     * @internal
+     */
+    public function isWritable(): bool;
```
Comment (Corepex) [inline models/DataObject/ClassDefinitionInterface.php]:
```suggestion
    
    /**
     * @internal
     */
    public function isWritable(): bool;
```

**Comment by Corepex** [comment]
@kjkooistra-youwe, could you please resolve the merge conflicts?

**Comment by Corepex** [comment]
@kjkooistra-youwe, I "cherry-picked" some of your changes. #15435

**Comment by kjkooistra-youwe** [comment]
> @kjkooistra-youwe, could you please resolve the merge conflicts?

Note: several previously updated classes were removed. These may have to be updated in external bundles.

- bundles/AdminBundle/src/Controller/Admin/DataObject/ClassController.php
- bundles/AdminBundle/src/Controller/Admin/DataObject/DataObjectActionsTrait.php
- bundles/AdminBundle/src/Controller/Admin/DataObject/DataObjectController.php
- bundles/AdminBundle/src/Controller/Admin/DataObject/DataObjectHelperController.php
- bundles/AdminBundle/src/Helper/GridHelperService.php
- bundles/EcommerceFrameworkBundle/tests/Support/Helper/EcommerceModel.php
- bundles/PersonalizationBundle/src/Document/Newsletter/AddressSourceAdapter/DefaultAdapter.php
- lib/Tool/Newsletter.php

File: models/DataObject/ClassDefinition/Helper/VarExportInterface.php [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+ *  @copyright  Copyright (c) Pimcore GmbH (http://www.pimcore.org)
+ *  @license    http://www.pimcore.org/license     GPLv3 and PCL
+ */
+
+namespace Pimcore\Model\DataObject\ClassDefinition\Helper;
+
+/**
+ * @internal
+ */
+interface VarExportInterface
```
Comment (dvesh3) [inline models/DataObject/ClassDefinition/Helper/VarExportInterface.php]: why not use the existing `Model\DataObject\ClassDefinition\Data\VarExporterInterface`?

**Review by dvesh3 — COMMENTED** [review]
Please resolve the conflicts.

File: models/DataObject/ClassDefinition/Helper/VarExportInterface.php (hunk 3 above)
Comment (dvesh3) [reply models/DataObject/ClassDefinition/Helper/VarExportInterface.php]: @kjkooistra-youwe please see :blush:

File: models/DataObject/ClassDefinition/Helper/VarExportInterface.php [hunk 4]
```diff
@@ -26,7 +26,7 @@ interface VarExportInterface
      * @return string[]
      */
     public function getBlockedVarsForExport(): array;
-
+
```
Comment (dvesh3) [inline models/DataObject/ClassDefinition/Helper/VarExportInterface.php]:
```suggestion

```

**Review by dvesh3 — APPROVED** [review]

**Comment by dvesh3** [comment]
@kjkooistra-youwe thanks a lot!

**Comment by fashxp** [comment]
@dvesh3 that is a proper BC break!! 
Especially models/DataObject/ClassDefinition.php must not change the method signatures!

**Comment by dvesh3** [comment]
@fashxp yes, we need to revert non-internal class changes. I will provide a follow up PR. thanks!

**Comment by dvesh3** [comment]
@kjkooistra-youwe we reverted non-internal class changes from this PR as these were leading to BC break. please see https://github.com/pimcore/data-hub/actions/runs/6092203292/job/16529960261

**Comment by kjkooistra-youwe** [comment]
@dvesh3 that pretty much reverts all the changes and still causes issues with mocking, for example. Will it be applied again in a future version? And if so, should a new issue be created to monitor it, similar to [#13595](https://github.com/pimcore/pimcore/issues/13595)?

**Comment by jdreesen** [comment]
> Will it be applied again in a future version?

Probably not until Pimcore 12... it's sad that they didn't manage to get it into Pimcore 11 :(

**Comment by dvesh3** [comment]
> @dvesh3 that pretty much reverts all the changes and still causes issues with mocking, for example. Will it be applied again in a future version? And if so, should a new issue be created to monitor it, similar to [#13595](https://github.com/pimcore/pimcore/issues/13595)?

yes, we need to come up with a strategy to not introduce BC breaks in minor releases. Otherwise these will be applied on Pimcore 12. Also have our eyes on #13595

**Comment by dvesh3** [comment]
> > Will it be applied again in a future version?
> 
> Probably not until Pimcore 12... it's sad that they didn't manage to get it into Pimcore 11 :(

@jdreesen yes, might be the case. Do you have an idea how to do it without introducing breaks?

**Comment by kjkooistra-youwe** [comment]
> > @dvesh3 that pretty much reverts all the changes and still causes issues with mocking, for example. Will it be applied again in a future version? And if so, should a new issue be created to monitor it, similar to [#13595](https://github.com/pimcore/pimcore/issues/13595)?
> 
> yes, we need to come up with a strategy to not introduce BC breaks in minor releases. Otherwise these will be applied on Pimcore 12. Also have our eyes on #13595

I linked it to that issue, since it's related. It's closed now though, so it might be a good idea to create a separate issue for the class definition.

**Comment by jdreesen** [comment]
> > > Will it be applied again in a future version?
> > 
> > 
> > Probably not until Pimcore 12... it's sad that they didn't manage to get it into Pimcore 11 :(
> 
> @jdreesen yes, might be the case. Do you have an idea how to do it without introducing breaks?

Unfortunately, no, I do not have one :/

**Pull request merged**